File: simple_sample/python/coroutine.py
import asyncio


async def test(name: str, n: float):
    print("start...", name)
    await asyncio.sleep(n)
    print("end...", name)
    return f"{name}:{n}"


async def main():
    result = await asyncio.gather(
        test("1番目", 2.5),
        test("2番目", 1.5),
        test("3番目", 3.0),
    )
    print(result)


if __name__ == "__main__":
    asyncio.run(main())

File: simple_sample/rb/coroutine.rb
# 2) async/await 風：Fiber＋簡易スケジューラ
def worker(name, delay_ms)
  Fiber.new do
    puts "#{name}: start"
    3.times do |i|
      puts "  #{name}: step #{i + 1}... start"
      Fiber.yield [:sleep, delay_ms / 1000.0]
      puts "  #{name}: step #{i + 1}... end"
    end
    puts "#{name}: done"
    :done
  end
end

def run(*fibers)
  queue = fibers.map { |f| [f, Time.now] }
  until queue.empty?
    queue.sort_by! { |(_, t)| t }
    fib, wake = queue.shift
    now = Time.now
    sleep(wake - now) if wake > now
    result = fib.resume
    if result.is_a?(Array)
      cmd, sec = result
      queue << [fib, Time.now + sec] if cmd == :sleep
    end
  end
  puts "all done"
end

puts "\n== Fiber-based coroutine demo =="
a = worker("A", 150)
b = worker("B", 120)
run(a, b)
